### PR TITLE
Week5 dev

### PR DESCRIPTION
**Pull request by vivianho18**
Given a meeting request with a list of attendees (including optional attendees) return the available times for the meeting. Includes relevant unit tests

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,84 @@
...
+
+
+  // Helper function to merge times ranges given in an ArrayList
+  // Returns an ArrayList of time ranges without any overlap 
+  private ArrayList<TimeRange> mergeTimes(ArrayList<TimeRange> ranges) {
+    ArrayList<TimeRange> mergedTimes = new ArrayList<>(); 
+    for (TimeRange range : ranges) {
+        if (mergedTimes.size() == 0) {
+            mergedTimes.add(range); 
+        } else if (!mergedTimes.get(mergedTimes.size() -1).overlaps(range)) {
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: `size() - 1` missing space

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,84 @@
...
+            mergedTimes.add(range); 
+        } else {
+            TimeRange oldRange = mergedTimes.remove(mergedTimes.size() - 1); 
+            int start = oldRange.start(); 
+            int end; 
+            if (oldRange.end() > range.end()) {
+                end = oldRange.end();
+            } else {
+                end = range.end();
+            }
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Alternatively `int end = Math.max(oldRange.end(), range.end())`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,84 @@
...
+    if (TimeRange.END_OF_DAY - startMeeting >= duration) {
+        TimeRange available = TimeRange.fromStartEnd(startMeeting, TimeRange.END_OF_DAY, true); 
+        availablities.add(available); 
+    }
+    return availablities; 
+  }
+
+
+  // Helper function to merge times ranges given in an ArrayList
+  // Returns an ArrayList of time ranges without any overlap
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You should note here that `ranges` needs to be sorted by start time (or sort ranges here instead)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,84 @@
...
+    ArrayList<TimeRange> unavailabilities = new ArrayList<>();
+    for (String attendee : attendees) {
+        for (Event event : events) {
+            for (String busyAttendee : event.getAttendees()) {
+                if (attendee.equals(busyAttendee)) {
+                    unavailabilities.add(event.getWhen()); 
+                }
+            }
+        }
+    }
```
Comment (ghost) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: Small optional optimization here:
You could make attendees into a HashSet or similar. Then you wouldn't need the outermost loop, as in the innermost loop you could do something like `if (attendees.contains(busyAttendee))`. This would be faster because HashSet lookup is constant time.

**Review by ghost — APPROVED** [review]

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java [hunk 5]
```diff
@@ -22,6 +22,6 @@
    * Returns a greeting for the given name.
    */
   public String greet(String name) {
-    return "Hello " + name;
+    return "Hello " + name.trim();
```
Comment (ghost) [inline walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java:25]: Also here's an example of where you could split this PR if you wanted to - you can split Greeter and GreeterTest into their own PR, as it's like a separate 'feature'. It's fine that the resulting PR would be very small.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (vivianho18) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: Thank you!
